group-coordinator/src/test/java/org/apache/kafka/coordinator/group/runtime/CoordinatorRuntimeTest.java

```diff
@@ -893,4 +1041,250 @@ public void testOnNewMetadataImage() {
+
+        // Timer #1.
+        ctx.coordinator.timer.schedule("timer-1", 10, TimeUnit.MILLISECONDS,
+            () -> Collections.singletonList("record1"));
+
+        // The coordinator timer should have one pending task.
+        assertEquals(1, ctx.timer.size());
+
+        // Advance time to fire the pending timer.
+        timer.advanceClock(10 + 1);
+
+        // An event should be waiting in the processor.
+        assertEquals(1, processor.size());
+
+        // Schedule a second timer with the same key.
+        ctx.coordinator.timer.schedule("timer-1", 10, TimeUnit.MILLISECONDS,
+            () -> Collections.singletonList("record2"));
+
+        // The coordinator timer should still have one pending task.
+        assertEquals(1, ctx.timer.size());
+
+        // Cancel the timer.
+        ctx.coordinator.timer.cancel("timer-1");
+
+        // The coordinator timer have no pending timers.
+        assertEquals(0, ctx.timer.size());
+
+        // Advance time to fire the cancelled timer.
+        timer.advanceClock(10 + 1);
+
+        // No new event expected because the timer was cancelled before
+        // it expired.
+        assertEquals(1, processor.size());
+
+        // Poll to execute the pending event.
+        assertTrue(processor.poll());
+
+        // Verify that no operation was executed.
+        assertEquals(Collections.emptySet(), ctx.coordinator.records());
+        assertEquals(0, ctx.timer.size());
+    }
+
+    @Test
+    public void testRetryTimer() throws InterruptedException {
+        MockTimer timer = new MockTimer();
+        CoordinatorRuntime<MockCoordinator, String> runtime =
+            new CoordinatorRuntime.Builder<MockCoordinator, String>()
+                .withTime(timer.time())
+                .withTimer(timer)
+                .withLoader(new MockCoordinatorLoader())
+                .withEventProcessor(new DirectEventProcessor())
+                .withPartitionWriter(new MockPartitionWriter())
+                .withCoordinatorBuilderSupplier(new MockCoordinatorBuilderSupplier())
+                .build();
+
+        // Loads the coordinator.
+        runtime.scheduleLoadOperation(TP, 10);
+
+        // Check initial state.
+        CoordinatorRuntime<MockCoordinator, String>.CoordinatorContext ctx = runtime.contextOrThrow(TP);
+        assertEquals(0, ctx.timer.size());
+
+        // Timer #1.
+        AtomicInteger cnt = new AtomicInteger(0);
+        ctx.coordinator.timer.schedule("timer-1", 10, TimeUnit.MILLISECONDS, () -> {
+            cnt.incrementAndGet();
+            throw new KafkaException("error");
+        });
+
+        // The coordinator timer should have one pending task.
+        assertEquals(1, ctx.timer.size());
+
+        // Advance time to fire the pending timer.
+        timer.advanceClock(10 + 1);
+
+        // The timer should have been called and the timer should have one pending task.
+        assertEquals(1, cnt.get());
+        assertEquals(1, ctx.timer.size());
```

Comment:
Every time we retry, we also push enqueue the new event right? Did we want to test this is handled correctly?

Reply:
I use the `DirectEventProcessor` in this test so we never have pending events in the queue. I did not test this here because previous tests already cover this.

Reply:
Sorry I misunderstood something about where the error was thrown, but I understand now.